File: src/main/python/AoC2016_20.py
#! /usr/bin/env python3
#
# Advent of Code 2016 Day 20
#

from __future__ import annotations
from typing import NamedTuple
from aoc import my_aocd
from functools import reduce


class Range(NamedTuple):
    minimum: int
    maximum: int

    @classmethod
    def between(cls, from_inclusive: int, to_inclusive: int) -> Range:
        return Range(from_inclusive, to_inclusive)

    def contains(self, element: int) -> bool:
        if element is None:
            return False
        return self.minimum <= element and element <= self.maximum

    def is_overlapped_by(self, other: Range) -> bool:
        if other is None:
            return False
        return other.contains(self.minimum) \
            or other.contains(self.maximum) \
            or self.contains(other.minimum)


def _parse(inputs: tuple[str]) -> list[Range]:
    return [Range.between(int(s[0]), int(s[1]))
            for s in [i.split('-') for i in inputs]]


def _combine_ranges(ranges: list[Range]) -> list[Range]:
    combines = list[Range]()
    ranges = sorted(ranges, key=lambda r: r.minimum)
    combines.append(ranges[0])
    for i in range(1, len(ranges)):
        r = ranges[i]
        combined = False
        for j, combine in enumerate(combines):
            if combine.is_overlapped_by(r):
                min_ = min(combine.minimum, r.minimum)
                max_ = max(combine.maximum, r.maximum)
                combines[j] = Range.between(min_, max_)
                combined = True
            elif r.maximum + 1 == combine.minimum:
                combines[j] = Range.between(r.minimum, combine.maximum)
                combined = True
            elif combine.maximum + 1 == r.minimum:
                combines[j] = Range.between(combine.minimum, r.maximum)
                combined = True
        if not combined:
            combines.append(r)
    return combines


def part_1(inputs: tuple[str]) -> int:
    combined = sorted(_combine_ranges(_parse(inputs)), key=lambda c: c.minimum)
    return combined[0].maximum + 1


def part_2(inputs: tuple[str]) -> int:
    combined = _combine_ranges(_parse(inputs))
    return reduce(lambda a, b: a - b,
                  map(lambda c: c.maximum - c.minimum + 1, combined),
                  2 ** 32)


TEST1 = '''\
5-8
0-2
4-7
'''.splitlines()
TEST2 = '''\
0-90000000
100000005-110000005
90000001-90000100
1000000054-1000000057
'''.splitlines()


def main() -> None:
    my_aocd.print_header(2016, 20)

    assert part_1(TEST1) == 3
    assert part_1(TEST2) == 90000101
    assert part_2(TEST1) == 4294967288
    assert part_2(TEST2) == 4194967190

    inputs = my_aocd.get_input(2016, 20, 1075)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")


if __name__ == '__main__':
    main()

File: src/main/java/AoC2016_20.java
import static java.util.Comparator.comparing;
import static java.util.stream.Collectors.toList;

import com.github.pareronia.aoc.RangeInclusive;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

@SuppressWarnings({"PMD.NoPackage", "PMD.ClassNamingConventions"})
public final class AoC2016_20 extends SolutionBase<List<RangeInclusive<Long>>, Long, Long> {

    private static final long TOTAL_IPS = 4_294_967_296L;

    private AoC2016_20(final boolean debug) {
        super(debug);
    }

    public static AoC2016_20 create() {
        return new AoC2016_20(false);
    }

    public static AoC2016_20 createDebug() {
        return new AoC2016_20(true);
    }

    @Override
    protected List<RangeInclusive<Long>> parseInput(final List<String> inputs) {
        final List<RangeInclusive<Long>> ranges =
                inputs.stream()
                        .map(s -> s.split("-"))
                        .map(sp -> RangeInclusive.between(Long.valueOf(sp[0]), Long.valueOf(sp[1])))
                        .collect(toList());
        return combineRanges(ranges);
    }

    // TODO: unduplicate w/ RangeInclusive.mergeRanges
    private List<RangeInclusive<Long>> combineRanges(final List<RangeInclusive<Long>> ranges) {
        final List<RangeInclusive<Long>> combines = new ArrayList<>();
        // why doesn't this work without sorting?
        ranges.sort(comparing(RangeInclusive::getMinimum));
        combines.add(ranges.get(0));
        for (int i = 1; i < ranges.size(); i++) {
            final RangeInclusive<Long> r = ranges.get(i);
            boolean combined = false;
            for (int j = 0; j < combines.size(); j++) {
                final RangeInclusive<Long> combine = combines.get(j);
                if (combine.isOverlappedBy(r)) {
                    final long min = Math.min(combine.getMinimum(), r.getMinimum());
                    final long max = Math.max(combine.getMaximum(), r.getMaximum());
                    combines.set(j, RangeInclusive.between(min, max));
                    combined = true;
                } else if (r.getMaximum() + 1 == combine.getMinimum()) {
                    combines.set(j, RangeInclusive.between(r.getMinimum(), combine.getMaximum()));
                    combined = true;
                } else if (combine.getMaximum() + 1 == r.getMinimum()) {
                    combines.set(j, RangeInclusive.between(combine.getMinimum(), r.getMaximum()));
                    combined = true;
                }
            }
            if (!combined) {
                combines.add(r);
            }
        }
        return Collections.unmodifiableList(combines);
    }

    @Override
    public Long solvePart1(final List<RangeInclusive<Long>> ranges) {
        return ranges.stream().findFirst().map(c -> c.getMaximum() + 1).orElseThrow();
    }

    @Override
    public Long solvePart2(final List<RangeInclusive<Long>> ranges) {
        return ranges.stream()
                .map(c -> c.getMaximum() - c.getMinimum() + 1)
                .reduce(TOTAL_IPS, (a, b) -> a - b);
    }

    @Samples({
        @Sample(method = "part1", input = TEST1, expected = "3"),
        @Sample(method = "part1", input = TEST2, expected = "90000101"),
        @Sample(method = "part2", input = TEST1, expected = "4294967288"),
        @Sample(method = "part2", input = TEST2, expected = "4194967190"),
    })
    public static void main(final String[] args) throws Exception {
        create().run();
    }

    private static final String TEST1 =
            """
            5-8
            0-2
            4-7
            """;
    private static final String TEST2 =
            """
            0-90000000
            100000005-110000005
            90000001-90000100
            1000000054-1000000057
            """;
}
